ArchiSteamFarm/Bot.cs

```diff
@@ -2836,6 +2839,295 @@ public sealed class Bot : IAsyncDisposable {
+
+				if (result.Count > 0) {
+					await Actions.SendInventory(result).ConfigureAwait(false);
+				}
+			} finally {
+				SendCompleteSetsSemaphore.Release();
+			}
+		}
+
+		private async Task<Dictionary<uint, byte>?> LoadCardsPerSet(ISet<uint> appIDs) {
+			if ((appIDs == null) || (appIDs.Count == 0)) {
+				throw new ArgumentNullException(nameof(appIDs));
+			}
+
+			switch (ASF.GlobalConfig?.OptimizationMode) {
+				case GlobalConfig.EOptimizationMode.MinMemoryUsage:
+					Dictionary<uint, byte> result = new Dictionary<uint, byte>(appIDs.Count);
+
+					foreach (uint appID in appIDs) {
+						byte cardCount = await ArchiWebHandler.GetCardCountForGame(appID).ConfigureAwait(false);
+
+						if (cardCount == 0) {
+							return null;
+						}
+
+						result.Add(appID, cardCount);
+					}
+
+					return result;
+				default:
+					IEnumerable<Task<(uint AppID, byte Cards)>> tasks = appIDs.Select(async appID => (AppID: appID, Cards: await ArchiWebHandler.GetCardCountForGame(appID).ConfigureAwait(false)));
+					IList<(uint AppID, byte Cards)> results = await Utilities.InParallel(tasks).ConfigureAwait(false);
+
+					if (results.Any(tuple => tuple.Cards == 0)) {
+						return null;
+					}
+
+					return results.ToDictionary(res => res.AppID, res => res.Cards);
+			}
+		}
+
+		internal static HashSet<Steam.Asset> GetItemsForFullBadge(IReadOnlyCollection<Steam.Asset> inventory, IReadOnlyDictionary<(uint RealAppID, Steam.Asset.EType Type, Steam.Asset.ERarity Rarity), (uint SetsToExtract, byte CardsPerSet)> amountsToExtract) {
+			if ((inventory == null) || (inventory.Count == 0) || (amountsToExtract == null) || (amountsToExtract.Count == 0)) {
+				throw new ArgumentNullException(nameof(inventory) + " || " + nameof(amountsToExtract));
+			}
+
+			HashSet<Steam.Asset> result = new HashSet<Steam.Asset>();
+			Dictionary<(uint RealAppID, Steam.Asset.EType Type, Steam.Asset.ERarity Rarity), Dictionary<ulong, List<Steam.Asset>>> itemsPerClassIDPerSet = inventory.GroupBy(item => (item.RealAppID, item.Type, item.Rarity)).ToDictionary(grouping => grouping.Key, grouping => grouping.GroupBy(item => item.ClassID).ToDictionary(group => group.Key, group => group.ToList()));
+
+			foreach (((uint RealAppID, Steam.Asset.EType Type, Steam.Asset.ERarity Rarity) set, (uint setsToExtract, byte cardsPerSet)) in amountsToExtract) {
+				if (!itemsPerClassIDPerSet.TryGetValue(set, out Dictionary<ulong, List<Steam.Asset>>? itemsPerClassID)) {
+					continue;
+				}
+
+				if (cardsPerSet < itemsPerClassID.Count) {
+					throw new ArgumentOutOfRangeException(nameof(inventory) + " && " + nameof(amountsToExtract));
+				}
+
+				if (cardsPerSet > itemsPerClassID.Count) {
+					continue;
+				}
+
+				foreach (List<Steam.Asset> itemsOfClass in itemsPerClassID.Values) {
+					uint classRemaining = setsToExtract;
+
+					for (int i = 0; (classRemaining > 0) && (i < itemsOfClass.Count); i++) {
```

Comment:
Since you use `itemsOfClass` for enumeration and linear access only, rewrite it into `foreach` loop with `classRemaining` condition as `break`, and change the type from `List` to `HashSet`.